{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1a0c5465-2944-48e2-93c6-27a2bfb0321c",
   "metadata": {},
   "source": [
    "# Downloading NEON Data\n",
    "NEON is a national network of research sites funded by the NSF. Each site has teams devoted to collecting a wide array of hydrology, biogeochemistry, and ecology data. A full list of the datasets available can be found at http://data.neonscience.org.\n",
    "\n",
    "We will be using the R package neonUtilities, which is designed to download and combine datasets the NEON data repository. We will start by downloading a few example datasets and then give you an opportunity to download your own datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f7484b16-4fa8-4fcf-9748-3c98bbffc363",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(neonUtilities)\n",
    "library(ggplot2)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1368aa42-c15a-4c23-85a0-b28ca7f2b95a",
   "metadata": {},
   "source": [
    "Jupyter Lab and Jupyter Notebook have a default setting for the R kernel that limits the numbers of columns that can be viewed. In the case of data from NEON and other data repositories with extensive metadata, the limitation on columns can make it difficul to view your data. To change this setting, you can use the function below. This will set the maximum number of columns and rows displayed. Here the columns displayed have been set to 40 and the rows displayed has been set to 10, but these can be changed as desired."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb40845b-17f8-4d5e-bc71-1a362ee69a6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.matrix.max.cols=40, repr.matrix.max.rows=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "057a5c09-8aca-4eda-a4bc-d7ec49aa4409",
   "metadata": {},
   "source": [
    "## Setting up your data storage space."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "386af484-57b2-4684-8add-84b81099378c",
   "metadata": {},
   "source": [
    "We need a place to save the files that you're going to download. When using bash and related terminals, the directory address _~/_ indicates the home directory for a given user. The command below creates a new directory in that home directory called _neon_data_. If the directory already exists, nothing will be changed and the _dir.create_ function will deliver a warning unless _showWarnings=FALSE_ is included in the command. We're doing this in the function below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d0632625-9c5e-4244-86fc-c5fdafc03e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(\"~/neon_data\", showWarnings=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a152362-2eea-4c70-802c-bba2899046d3",
   "metadata": {},
   "source": [
    "## Downloading hydrology data\n",
    "We will start by downloading hydrology data including discharge, velocity, and channel geometry parameters. We will begin with discharge. Discharge. Each data product at NEON has its own unique [site](https://www.neonscience.org/field-sites/explore-field-sites) and [data](https://data.neonscience.org/data-products/explore) identifier codes that we use to specify the dataset we want to download in the code below.\n",
    "\n",
    "We will begin by downloading hydrology data for the Arikee River (code ARIK) from the \"Continuous Discharge\" Data Product (DP4.00130.001).\n",
    "\n",
    "We will use the function _loadByProduct()_ to download the data. If you have any questions about a given function, you can type _?functionName_, in this case _?loadByProduct_ into a code cell and press ctrl+Enter (or cmd+Enter for macOS users) to get help documentation for that function.\n",
    "\n",
    "### Discharge\n",
    "\n",
    "We will begin by loading discharge data from the Arikee River into the data object _Arik_Q_. We use the option _chec.size = FALSE_, which removes an interactive step that requires confirmation download confirmation downloading the requested data. NEON datasets are rarely more than 500 mb in size. If such datasets may cause space concerns, you can change the option to _check.size = TRUE_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82a3ec4a-c478-4a0e-a15b-2a676c74d680",
   "metadata": {},
   "outputs": [],
   "source": [
    "?loadByProduct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9f6330cd-3bd0-4f9f-aa20-a151d4efb9a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Finding available files\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Provisional data were excluded from available files list. To download provisional data, use input parameter include.provisional=TRUE.\n",
      "\n",
      "Downloading files totaling approximately 49.510273 MB\n",
      "\n",
      "Downloading 9 files\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Unpacking zip files using 1 cores.\n",
      "\n",
      "Stacking data files\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Finished: Stacked 2 data tables and 4 metadata tables!\n",
      "\n",
      "Stacking took 7.610836 secs\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Arik_Q<-loadByProduct(dpID = \"DP4.00130.001\",site = \"ARIK\", check.size = FALSE, startdate = \"2023-01\", enddate = \"2023-12\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a423cd2e-834a-4deb-9dca-ce1e8e833405",
   "metadata": {},
   "source": [
    "The dataset you've downloaded contains continuous discharge data as well as a large amount of ancillary metadata. To start unpacking your data, use the _names_ function in R.\n",
    "\n",
    "Note that there is a lot of data here. This includes a lot of useful metadata, but we just want to look at the discharge estimates right now. See if you can figure out from the results of the structure command the location of the continuous discharge data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d43ee249-7493-4986-877f-5504dfae49bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(Arik_Q)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b2cb806-aa85-4521-a707-15554b210a42",
   "metadata": {},
   "source": [
    "You can check the contents of each of these list elements by using the command _head_ and include the name of the list _Arik_Q_ and the names of the list element that you want to view followed by a dollar sign. The command _head_ will show the first six rows of the data frame or matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd803d8f-31da-4516-99dc-338d6d5f64e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(Arik_Q$csd_continuousDischarge)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f9199f3-574a-44d3-a9ed-d897e4bcd73c",
   "metadata": {},
   "source": [
    "The discharge time series is inclued in the dataframe \"csd_continuousDischarge\". Within this dataframe there are a number of columns, the one with the discharge data is called _maxpostDischarge_. The structure of online datasets from NEON and data repositories can sometimes be unintuitive. Fortunately, online repositories also include documentation. This includes a [metadata document from NEON](file:///Users/jdh/Downloads/NEON_continuousQ_userGuide_vD.1.pdf) that describes the nature of data product DP4.00130.001, continuous discharge,in detail. These can be found associated with each dataproduct at data.neonscience.org."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12470d53-3542-42a5-8ca2-34d1c3069a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_Q_cD<-Arik_Q[[\"csd_continuousDischarge\"]]\n",
    "head(Arik_Q_cD)\n",
    "names(Arik_Q_cD)\n",
    "Arik_Q_cD$maxpostDischarge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26a311ca-5b34-45ea-83cb-47bb182c28c3",
   "metadata": {},
   "source": [
    "Now that we've found the dataset, let's plot it using the ggplot package. Fortunately, this dataset includes a formatted date/time column so plotting is easy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b439762f-b7dd-49b0-a295-ca61f0dbf632",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(Arik_Q_cD,aes(endDate,maxpostDischarge))+\n",
    "geom_point(size=2)+\n",
    "geom_line()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589cbb98-89cd-4793-8829-592bbcb6ea28",
   "metadata": {},
   "source": [
    "#### Try downloading from another site of your choice. Save the continuous discharge dataset in the same directory as we saved the previous file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a068fd35-3e9c-4a78-8c37-e9ebe1c60862",
   "metadata": {},
   "source": [
    "### Channel Geometry\n",
    "Channel geometry is a useful parameter for ecological analysis and some of the analyses that we will conduct in subsequent modules. We can use NEON's [Discharge Field Collection (DP1.20048.001)](https://data.neonscience.org/data-products/DP1.20048.001) to obtain these measurements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f9314ce-30a4-4a29-86b3-21798ffa1918",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_DFC<-loadByProduct(dpID = \"DP1.20048.001\",site = \"ARIK\", check.size = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5323610-da41-47f4-9bce-00895a2ccf18",
   "metadata": {},
   "source": [
    "As before, we have to use the structure command to find the data we need. If you review the NEON documentation for this dataset, you will note that field surveys are conducted with the aid of an accoustic doppler current profiler (ADCP). The results of these surveys are summarized in the  list element _dsc_individualFieldData. This is the dataframe that has the hydraulic geometry parameters we're looking for!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39db1c89-9d8b-49eb-8b40-88c751126224",
   "metadata": {},
   "outputs": [],
   "source": [
    "str(Arik_DFC)\n",
    "Arik_HG<-Arik_DFC$dsc_fieldDataADCP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa0330ab-d304-4fe5-b84d-ef52e270404a",
   "metadata": {},
   "source": [
    "We can use these data to assess the relationship between different hydraulic parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "334ffec5-b36c-4049-9f05-73c453c0d0a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(Arik_HG)\n",
    "head(Arik_HG)\n",
    "Arik_HG$Channel_Width<-Arik_HG$sectionArea/Arik_HG$waterDepth\n",
    "ggplot(Arik_HG,aes(waterDepth,sectionArea))+geom_point()\n",
    "ggplot(Arik_HG,aes(Channel_Width,waterDepth))+geom_point()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60234532-0e2a-472b-a35d-106a0ed0441d",
   "metadata": {},
   "source": [
    "### Rating Curve\n",
    "Finally, we want to be able to download, view, and evaluate rating curves that were used to produce continuous discharge measurements from pressure transducers. These are obtained from the [Stage-Discharge Rating Curves (DP4.00133.001)](https://data.neonscience.org/data-products/DP4.00133.001) dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb480c8-1781-4cc4-84c0-2230f7e6fcfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_RC<-loadByProduct(dpID = \"DP4.00133.001\",site = \"ARIK\", check.size = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cdeb5e2-0c7f-49a6-8957-3ae054cbff62",
   "metadata": {},
   "outputs": [],
   "source": [
    "str(Arik_RC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94f888a6-a0a4-491d-adb6-35dd90c4bfba",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_DC<-Arik_RC$sdrc_gaugeDischargeMeas\n",
    "names(Arik_DC)\n",
    "str(Arik_DC)\n",
    "head(Arik_DC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65c7730c-f5ad-42b1-bbed-2ba92913eda4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(Arik_DC,aes(gaugeHeight,streamDischarge))+geom_point()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22b3506c-2be8-4b3c-9216-3e2464038df2",
   "metadata": {},
   "source": [
    "## Water Chemistry Data\n",
    "There are two types of water chemistry data available from NEON and other portals. There are individually collected grab samples as well as continuously collected sensor data. Both types of data have their own advantages and drawbacks and are most useful when collected side-by-side. NEON collects a variety of grab samples and also deploys a number of _in situ_ water sensors.\n",
    "\n",
    "Grab samples collected include [water quality/chemistry (anions, cations, conductivity, pH, and various forms of carbon and\n",
    "nutrients including total, dissolved and particulates)](https://data.neonscience.org/data-products/DP1.20093.001), dissolved gasses (nitrous oxide, methane, carbon dioxide), and surface water stable isotopes ()\n",
    "\n",
    "Sensors deployed for _in situ_ data collection include a [YSI EXO 2 Multiparameter Sonde (temperature, conductance, pH, fluorescence dissolved organic matter (fDOM), dissolved oxygen)](https://data.neonscience.org/data-products/DP1.20288.001), [SUNA V2 Nitrate Sensor (nitrate, absorbance at 254 nm)](https://data.neonscience.org/data-products/DP1.20033.001).\n",
    "\n",
    "### Grab Sample Data\n",
    "We will download grab sample data, again using the Arikee River site as our example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d983c2-3683-4144-a886-5bdd37ead4fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_WC<-loadByProduct(dpID = \"DP1.20093.001\",site = \"ARIK\", check.size = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e77c7f51-e66a-4ca8-9013-b2877a33d83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "str(Arik_WC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12437058-c9bc-4627-bf7d-a5e90bd996fe",
   "metadata": {},
   "source": [
    "Let's extract nitrate, chloride, and pH values and plot these over time!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27c7f17a-2fae-4092-9eb3-7a5c4e83c6de",
   "metadata": {},
   "source": [
    "There are two datasets of interest, the first contains data collected by the local NEON domain (swc_domainLabData), the second contains data collected by an externally contracted lab (swc_externalLabDataByAnalyte), we're going to extract both of these datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d90437a9-2187-499e-9916-5969993dd0a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_DomainWC<-Arik_WC$swc_domainLabData\n",
    "Arik_ExternalWC<-Arik_WC$swc_externalLabDataByAnalyte"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8374251d-0799-4a8b-a51f-b6b201663d61",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(Arik_DomainWC)\n",
    "head(Arik_DomainWC)\n",
    "names(Arik_ExternalWC)\n",
    "head(Arik_ExternalWC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c38ea177-4e7c-4c2c-8786-5b7533962d2e",
   "metadata": {},
   "source": [
    "As you can see, most of the water chemistry data are contained in the Arik_ExternalWC dataset. You can examine the analyte for each dataset by accessing the columns as shown in the cell below. Let's plot iron (Fe) and Nitrate/Nitrate (NO3+NO2 - N) values over time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaa6b0a4-595d-4a37-a895-1d29e2ae425e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "unique(Arik_ExternalWC$analyte)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34091768-e779-4c3a-9eb6-85243af19ec2",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_WC_Plot<-subset(Arik_ExternalWC,analyte==\"Fe\" | analyte ==\"NO3+NO2 - N\")\n",
    "ggplot(Arik_WC_Plot,aes(x=collectDate,y=analyteConcentration,color=analyte))+\n",
    "    geom_point(size=2)+\n",
    "    facet_wrap(.~analyte,ncol=1,scales=\"free_y\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bd5d7b6-3590-4ecd-9956-a46e4bf2d5d4",
   "metadata": {},
   "source": [
    "### Photosynthetically Active Radiation Data\n",
    "Photosynthetically active radiation (PAR) is a key input for many ecosystem models that involve photosynthesis. Given the importance of these data to future models, we will learn how to download these data here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91647b30-dd3a-4ff2-a1c4-cabdb71f97c4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e77acc59-1dbc-4c2e-845c-52a047bcae60",
   "metadata": {},
   "source": [
    "### Continuous Sensor Data\n",
    "Now we will practice downloading, extracting, and plotting _in situ_ sensor data using the Water Quality (DP1.20288.001) dataset. First we will download the dataset and examine its structure. For the YSI EXO 2 water quality data, the dataframe is contained in the list as \"waq_instantaneous\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51abb105-3977-4e34-ab5d-3864801eaf14",
   "metadata": {},
   "outputs": [],
   "source": [
    "Arik_WQ<-loadByProduct(dpID = \"DP1.20288.001\",site = \"ARIK\", check.size = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e34efa8d-18d7-44c7-9f0f-c7de03a6bb17",
   "metadata": {},
   "outputs": [],
   "source": [
    "str(Arik_WQ)\n",
    "Arik_WQ_Data<-Arik_WQ$waq_instantaneous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79eb0b7f-1502-4fda-9ed5-341affcc2366",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(Arik_WQ_Data)\n",
    "names(Arik_WQ_Data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fafb57c-c0e1-4758-903b-b5e84a126eeb",
   "metadata": {},
   "source": [
    "Now we can easily plot different parameters using ggplot or the plotting function of choice. In this example, we are plotting specific conductance (specificConductance) and dissolved oxygen percent saturation (localDissolvedOxygenSat). Remember, as always, you can reference the data product documentation if you are unsure about the meaning of a given variable. Caution, the dataset is large so if you plot the entire dataset this may take a moment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b308351-e734-4ef6-adf1-41a4b25bc174",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot(Arik_WQ_Data,aes(x=startDateTime,y=specificConductance))+\n",
    "    geom_point()\n",
    "\n",
    "ggplot(Arik_WQ_Data,aes(x=startDateTime,y=localDissolvedOxygenSat))+\n",
    "    geom_point()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97f2d6a5-07d2-4f84-aeba-8cd1dfb26373",
   "metadata": {},
   "source": [
    "## Biodiversity Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e2c61f4-74c0-463b-afdc-74207bcd0d41",
   "metadata": {},
   "source": [
    "### Macroinvertebrate Biodiversity Data\n",
    "Macroinvertebrate diversity is a key dataset for water quality regulations. Many of the determinations regarding water quality under federal water protection laws in the United States are based on the diversity of benthic macroinvertebrates in streams and rivers.\n",
    "\n",
    "[Macroinvertebrate collection](https://data.neonscience.org/data-products/DP1.20120.001) is conducted routinely at all NEON aquatic sites. We are going to download and extract macroinvertebrate species counts in preparation for "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1cec0a91-65c6-4336-976c-d178cb125b0d",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "expression cannot contain assignment, perhaps you meant \"==\"? (1752570270.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[2], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    ARIK_MI<-loadByProduct(dpID = \"DP1.20120.001\",site = \"ARIK\", check.size = FALSE)\u001b[0m\n\u001b[0m                                                                 ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m expression cannot contain assignment, perhaps you meant \"==\"?\n"
     ]
    }
   ],
   "source": [
    "ARIK_MI<-loadByProduct(dpID = \"DP1.20120.001\",site = \"ARIK\", check.size = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d43c08e1-3f61-4d25-b726-9c2afe65c6f9",
   "metadata": {},
   "source": [
    "Again we want to look at the names of the dataset to see where the macroinvertebrate biodiversity data are located. Note the list element _inv_taxonomyProcessed_, this contains the processed taxonomy that you want to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e23ff0f9-b671-4e57-a4c8-2e58d65b53e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(ARIK_MI)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fa7a36a-2889-427d-b68a-7768190609e6",
   "metadata": {},
   "source": [
    "As before we want to save the processed taxonomy data in a new dataframe so that it can easily be accessed in the future."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48c7cf18-391e-43ff-a289-129b981c2085",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARIK_MI_Diversity<-as.data.frame(ARIK_MI$inv_taxonomyProcessed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80af38fa-4a10-4935-a5e7-fbfa37decf95",
   "metadata": {},
   "source": [
    "There are many variables included in this dataset. We are particularly interested in the macroinvertebrate taxonomy. This taxonomy information includes _phylum_, _subphylum_, _class_, _subclass_, _infraclass_, _superorder_, _order_, _suborder_, _infraorder_, _superfamily_, _family_, _subfamily_, _tribe_, _subtribe_, and _genus_. In addition, _individualCount_ is a useful variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51b013fe-abe0-4a53-aa21-a286a7257075",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(ARIK_MI_Diversity)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fc64f57-1e53-4913-83a1-898504f7c441",
   "metadata": {},
   "source": [
    "We are going to select the important columns before we save this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02fb611f-f2d3-4fee-9689-c620b851de6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARIK_MI_Taxonomy<-ARIK_MI_Diversity[,c('phylum','subphylum','class','subclass','infraclass','superorder','order','suborder','infraorder','superfamily','family','subfamily','tribe','subtribe','genus','individualCount')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "514e2462-3c47-4623-9f45-c7a48b22550b",
   "metadata": {},
   "source": [
    "Now we're going to save our dataset as an rds file as before. We want to keep our data organized, so we're storing it in a new sub-directory of your _neon_data_ directory called _biodiversity_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc65bdf-4c40-46b0-ae85-80e065bf9b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(\"~/neon_data/biodiversity\",showWarnings=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "194e040e-f057-4682-b9c2-88a08753a6fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "Then we can save the RDS file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69264017-d916-4671-a6af-1b17f0751929",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(ARIK_MI_Taxonomy,\"~/neon_data/biodiversity/ARIK_MI_Taxonomy.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93de9fe8-e9e7-49df-b73e-88ecf19080ac",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "source": [
    "### Fish Diversity Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c1710b4-1801-4776-817f-6fc310c96903",
   "metadata": {},
   "source": [
    "Now we want to download some fish diversity data. Referencing the [data codes](https://data.neonscience.org/data-products/explore), fish diversity data from field surveys are found in [DP1.20107.001](https://data.neonscience.org/data-products/DP1.20107.001). We will again download data from the ARIK site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f596d00-981f-4fdc-b529-d9a2b21c5970",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ARIK_Fish<-loadByProduct(dpID = \"DP1.20107.001\",site = \"ARIK\", check.size = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa9b8cbc-28d9-4874-a598-f95389acda10",
   "metadata": {},
   "source": [
    "Now let's look at the structure of the fish dataset we just downloaded using the _names_ funciton as we've done before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "000ab5e5-3696-4068-bc48-2f400f3fcc09",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(ARIK_Fish)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd7b6eb2-d8cc-48a6-8647-2ad7f429949f",
   "metadata": {},
   "source": [
    "Again we see a few possible options, but here _fsh_bulkCount_ is the "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab94ab19-1fc7-473d-9923-f01cd543d5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARIK_Fish_Taxonomy<-as.data.frame(ARIK_Fish$fsh_bulkCount)\n",
    "names(ARIK_Fish_Taxonomy)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e7a7836-c051-4f28-93b6-5c4306f57e8a",
   "metadata": {},
   "source": [
    "Now let's take a look at the dataset of fish counts. Note that the format of this file is different and fish scientific names (genus and species only) are report as a single column rather than the entire taxonomy of each species with each attribute as a separate column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa64ddc4-5407-4466-b3a4-b0bb4960b34f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ARIK_Fish_Taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d27f4d44-5fc9-44d5-baf0-cbeea52cff57",
   "metadata": {},
   "source": [
    "The code below will help quickly review the data you've downloaded. There are multiple rows with the same data in this dataset, so the first part of the code block below uses functions from dplyr to sum up the number of fish by scientific name. After that we plot the counts using a bar graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1c32024-f37d-4117-8046-3244e1e8ac0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "arik_fish_data_plot <- ARIK_Fish_Taxonomy %>%\n",
    "  group_by(scientificName) %>%\n",
    "  summarize(Total_Count = sum(bulkFishCount))\n",
    "\n",
    "# Create a bar plot using ggplot2\n",
    "ggplot(arik_fish_data_plot, aes(x = scientificName, y = Total_Count)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  labs(title = \"Species Counts\", x = \"Species\", y = \"Count\") +\n",
    "  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3d051c2-5c96-44d9-8bec-dddff2335f82",
   "metadata": {},
   "source": [
    "Finally, we will save the taxonomic diversity dataframe as a RDS file as well have done before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "992f60ea-dda7-4e5e-8722-a03a03ced0d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(ARIK_Fish_Taxonomy,\"~/neon_data/biodiversity/ARIK_Fish_Taxonomy.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5180c4a2-3771-4de4-b347-a80df1fe0d0d",
   "metadata": {},
   "source": [
    "### Microbial "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "767376b0-4006-4b06-be52-25b54007eb5f",
   "metadata": {},
   "source": [
    "Now we want to download some microbial diversity data. Referencing the [data codes](https://data.neonscience.org/data-products/explore), aquatic benthic microbial diversity data from field surveys are found in [DP1.20086.001](https://data.neonscience.org/data-products/DP1.20086.001). Abundance data includes archaea, bacteria, and fungi. We will again download data from the ARIK site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0932e50-297f-4232-9882-412a22b1659c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARIK_Microbe<-loadByProduct(dpID = \"DP1.20086.001\",site = \"ARIK\", check.size = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d67831bd-7f16-4978-aae9-e87baddb4e7c",
   "metadata": {},
   "source": [
    "Looking at the names of the different list elements in the downloaded microbial community composition dataset, there are two sents of sequences, _mcc_benthicSeqVariantMetadata_16S_, which contains sequences for prokaryotes and _mcc_benthicSeqVariantMetadata_ITS_, which contains sequences for fungi. Each of these datasets includes samples from many different substrates including sand (episammon) and benthic plants (epiphyton).\n",
    "\n",
    "In this case these two dataframes (_mcc_benthicSeqVariantMetadata_16S_ and _mcc_benthicSeqVariantMetadata_ITS_) do not contain the actual abundance data, but instead contain urls to csvs containing the relevant datasets for each individual sample. We're going to take a look at both of these datasets and then we will use a script to download all the CSVs and combine them into a single dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da138da7-9900-4e79-984a-088969f2c4bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(ARIK_Microbe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4baff3c4-72fa-4751-a2ea-00b58cc851a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARIK_Microbe$mcc_benthicSeqVariantMetadata_16S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e63ce387-a8e2-4ab0-9ba7-989975702105",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARIK_Microbe$mcc_benthicSeqVariantMetadata_ITS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d79962f-f3b7-46c2-998a-ca96aab786c9",
   "metadata": {},
   "source": [
    "Now we're using a the urls to combine all the data into a single dataframe. We will do this separate for the 16S and ITS dataframes and then combine those."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42111461-391f-496e-97c4-66af54ef51e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#downloadFileUrl\n",
    "#geneticSampleID\n",
    "read_csv_filename <- function(filename,dataname){\n",
    "    ret <- read.csv(filename)\n",
    "    ret$Source <- filename\n",
    "    return(ret)\n",
    "}\n",
    "ARIK_bcc_urls<-ARIK_Microbe$mcc_benthicSeqVariantMetadata_16S\n",
    "ARIK_fcc_urls<-ARIK_Microbe$mcc_benthicSeqVariantMetadata_ITS\n",
    "\n",
    "\n",
    "combined_data_16s <- ARIK_bcc_urls$downloadFileUrl %>%\n",
    "    lapply(read_csv_filename) %>%\n",
    "    bind_rows\n",
    "\n",
    "combined_data_ITS <- ARIK_fcc_urls$downloadFileUrl %>%\n",
    "    lapply(read_csv_filename) %>%\n",
    "    bind_rows\n",
    "\n",
    "ARIK_combined_microbe_data<-bind_rows(combined_data_16s,combined_data_ITS)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fd905ab-969f-48e6-9688-c5fd3be4c0ae",
   "metadata": {},
   "source": [
    "Now let's take a look at the dataset that we've compiled and assembled:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5be285b8-d097-49db-8d1e-a5b90a71f96d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARIK_combined_microbe_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b29527db-a0ad-46d8-9eb1-57bd1d2d7953",
   "metadata": {},
   "source": [
    "One thing you may notice is that this dataset is much larger than previous datasets. This is due to the understandably high individual count and species abundance of microbes in the environment. Also note that the taxonomic information is again formatted differently in this dataset. Here the variable _completeTaxonomy_ includes the entire taxonomy of an organisms from domain down to species, if that has been identified with _;_ separating the different levels of taxonomy. Each level of taxonomic identification is also included in its own column.\n",
    "\n",
    "We will learn how to plot and analyze these data in future modules, so here we will only display the number of rows in the dataset, which indicates how complex the diversity of microbial samples are, alongside the unique number of operational taxonomic units (OTUs). The OTU concept is used in microbial ecology because the concept of a species in microbial diversity, particularly as it relates to archaea and prokaryotes, is difficult to define due to asexual reproduction and the prevalence of horizontal gene transfer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42a99bfc-ba16-4c1d-ac2d-27a989430c36",
   "metadata": {},
   "outputs": [],
   "source": [
    "paste(\"Number of rows of dataset:\",nrow(ARIK_combined_microbe_data))\n",
    "paste(\"Number of unique OTUs:\",length(unique(ARIK_combined_microbe_data$completeTaxonomy)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e61fd3-f96d-4f97-9002-bf6a1a61cdc8",
   "metadata": {},
   "source": [
    "Finally we're going to save the microbial diversity dataset for future use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3656256-a0c2-4606-a3d4-bae5fa27ae06",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(ARIK_combined_microbe_data,\"~/neon_data/biodiversity/ARIK_combined_microbe_data.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e39a95f-ccfb-48f8-8669-7701289a3028",
   "metadata": {},
   "source": [
    "## Assignment\n",
    "### Now it's your turn to download some data!\n",
    "For each of the data groups above, download datasets from two different NEON sites. For each of the following data groups, do the following to display the contents of the datasets you've downloaded:\n",
    "* Present a graph o one year of data for each of the following data types (you choose the year):\n",
    "    * Discharge\n",
    "    * Channel Geometry\n",
    "    * Rating Curve\n",
    "    * Water quality sonde continuous data\n",
    "    * Photosynthetically Active Radiation (PAR)\n",
    "* For the following diversity metrics, down the data and present a figure with species taxonomic Order by counts:\n",
    "    * Fish Biodiversity\n",
    "    * Macroinvertebrate Biodiversity\n",
    "* For microbial diversity, present a count of the unique number of OTUs present in your dataset:\n",
    "    * Microbe Biodiversity"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
